gherkin/php/src/AstNode.php

```diff
@@ -0,0 +1,99 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Cucumber\Gherkin;
+
+use Cucumber\Gherkin\Parser\RuleType;
+use Cucumber\Gherkin\Parser\TokenType;
+use LogicException;
+
+final class AstNode
+{
+    /**
+     * @var array<string, list<object|list<object>|string>>
+     */
+    private array $subItems = [];
+
+    public function __construct(
+        public readonly RuleType $ruleType
+    ) {
+    }
+
+    /**
+     * @param object|list<object>|string $entry
+     */
+    public function add(RuleType $ruleType, object|array|string $entry): void
+    {
+        $this->subItems[$ruleType->name] ??= [];
+        $this->subItems[$ruleType->name][] = $entry;
+    }
+
+    /**
+     * @template T of object
+     *
+     * @param class-string<T> $expectedType
+     *
+     * @psalm-return list<T>
+     */
+    public function getItems(string $expectedType, RuleType $ruleType): array
+    {
+        $items = $this->subItems[$ruleType->name] ?? [];
+
+        /**
+         * Force the type because we trust the parser, could be validated instead
+         * @var list<T> $items
+         */
+
+        return $items;
+    }
+
+    /**
+     * @template S of object
+     *
+     * @param class-string<S> $expectedType
+     * @param S|null $defaultValue
+     *
+     * @psalm-return S|null
```

Comment:
I only returns null if $defaultValue is null, so this can probably be a conditional type annotation and remove some null checks elsewhere